### ISSUE TITLE
Add support for runtime styling when snapshotting

### ISSUE THREAD

**Issue by frederoni** [mapbox/mapbox-gl-native#9914]
Now that snapshots landed in Android/Core #9748 & iOS/macOS #9891, let's take it one step further by exposing runtime styling to the snapshotters:

On the iOS/macOS side, I'd propose something along these lines:
```objc
[snapshotter startWithStyleHandler:^(MGLStyle *style) {
    [style addLayer:polyline];
    [style addSource:source];
} completionHandler:^(UIImage *snapshot, NSError *error) {
    
}];
```

@ivovandongen @fabian-guerra @tobrun

**Comment by 1ec5** [comment]
On iOS/macOS, would chaining operations to the snapshot’s start require something along the lines of #6386?

**Comment by frederoni** [comment]
@1ec5 #6386 would probably come in handy. I figured modifying the style later would make more sense when being used in a scroll view instead of allocating a map view and style upon initialization but I don't have any strong opinion about that so far.

**Comment by 1ec5** [comment]
If this feature is implemented, we’d be able to conditionally enable transparency in snapshots (#10118) based on whether the background style layer is transparent.

**Comment by 1ec5** [comment]
Fixed in #16268. mapbox/mapbox-gl-native-android#220 tracks hooking the feature up on Android. mapbox/mapbox-gl-native-ios#200 tracks hooking the feature up on iOS and macOS.
